src/graphs.rs

```diff
@@ -0,0 +1,312 @@
+  root_packages: HashMap<LockfilePkgReq, LockfilePkgId>,
+  packages: HashMap<LockfilePkgId, LockfileGraphPackage>,
+  remotes: BTreeMap<String, String>,
+  nv_to_jsr_url: FNvToJsrUrl,
+}
+
+impl<FNvToJsrUrl: Fn(&str) -> Option<String>>
+  LockfilePackageGraph<FNvToJsrUrl>
+{
+  pub fn from_lockfile<'a>(
+    content: PackagesContent,
+    remotes: BTreeMap<String, String>,
+    old_config_file_packages: impl Iterator<Item = &'a str>,
+    nv_to_jsr_url: FNvToJsrUrl,
+  ) -> Self {
+    let mut root_packages =
+      HashMap::<LockfilePkgReq, LockfilePkgId>::with_capacity(
+        content.specifiers.len(),
+      );
+    // collect the specifiers to version mappings
+    let mut packages = HashMap::new();
+    for (key, value) in content.specifiers {
+      if let Some(value) = value.strip_prefix("npm:") {
+        root_packages.insert(
+          LockfilePkgReq(key.to_string()),
+          LockfilePkgId::Npm(LockfileNpmPackageId(value.to_string())),
+        );
+      } else if let Some(value) = value.strip_prefix("jsr:") {
+        let nv = LockfilePkgId::Jsr(LockfileJsrPkgNv(value.to_string()));
+        root_packages.insert(LockfilePkgReq(key), nv.clone());
+        packages.insert(
+          nv,
+          LockfileGraphPackage::Jsr(LockfileJsrGraphPackage::default()),
+        );
+      }
+    }
+
+    for (nv, content_package) in content.jsr {
+      let new_deps = &content_package.dependencies;
+      let package = packages
+        .entry(LockfilePkgId::Jsr(LockfileJsrPkgNv(nv.clone())))
+        .or_insert_with(|| {
+          LockfileGraphPackage::Jsr(LockfileJsrGraphPackage::default())
+        });
+      match package {
+        LockfileGraphPackage::Jsr(package) => {
+          package.dependencies = new_deps
+            .iter()
+            .map(|req| LockfilePkgReq(req.clone()))
+            .collect();
+        }
+        LockfileGraphPackage::Npm(_) => unreachable!(),
+      }
+    }
+    for (id, package) in content.npm {
+      packages.insert(
+        LockfilePkgId::Npm(LockfileNpmPackageId(id.clone())),
+        LockfileGraphPackage::Npm(LockfileNpmGraphPackage {
+          root_ids: Default::default(),
+          integrity: package.integrity.clone(),
+          dependencies: package
+            .dependencies
+            .iter()
+            .map(|(key, dep_id)| {
+              (key.clone(), LockfileNpmPackageId(dep_id.clone()))
+            })
+            .collect(),
+        }),
+      );
+    }
+
+    let mut root_ids = old_config_file_packages
+      .filter_map(|value| {
+        root_packages
+          .get(&LockfilePkgReq(value.to_string()))
+          .cloned()
+      })
+      .collect::<Vec<_>>();
+
+    // trace every root identifier through the graph finding all corresponding packages
+    while let Some(root_id) = root_ids.pop() {
+      let mut pending = VecDeque::from([root_id.clone()]);
+      while let Some(id) = pending.pop_back() {
+        if let Some(package) = packages.get_mut(&id) {
```

Comment:
What if `pacakges.get_mut()` returns `None`? Is it assume it will be found by another root?

Reply:
It might occur if someone has been modifying the lockfile themselves. It's not a big deal in that case because then the package is already gone.